Visual improvements to CoverageTrack

danvk opened the pull request:
This includes a bunch of changes:

1. eliminates the Moiré pattern (fixes #334). This required splitting out the mismatch rendering code into a separate loop.
1. rounds the y-axis coordinates to get crisper rendering. IGV.js does this and it looks much nicer.
1. fixes a bug where the bars were all ~10px too high, thus giving incorrect counts. Funny that we never noticed this!
1. restored the white border on the coverage track scale. I believe this was lost in the SVG→canvas transition (we used to do it with CSS).

Here's a before/after:

![screen shot 2015-10-27 at 1 20 59 pm](https://cloud.githubusercontent.com/assets/98301/10766375/9e3f98c8-7cad-11e5-967a-eaa61fe172f3.png)

![screen shot 2015-10-27 at 1 20 14 pm](https://cloud.githubusercontent.com/assets/98301/10766372/9be15508-7cad-11e5-9469-13e350e61361.png)

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/hammerlab/pileup.js/338)
<!-- Reviewable:end -->

armish commented:
Thanks for all these fixes, @danvk! Sorry about the 10px-bug, probably introduced it while transitioning to canvas :\

LGTM!
<!-- Reviewable comment -K1e_sHhfTPaF7VgRrcL:-634324563 -->

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

<sup>**[src/main/CoverageTrack.js, line 174 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/338#-K1eYxzEdqaB2ozcA14i)** ([raw file](https://github.com/hammerlab/pileup.js/blob/a15c96995339285e5cab206b984f2d9c4cdfb29e/src/main/CoverageTrack.js#L174)):</sup>
why do you have those parentheses around this statement?
<!-- Reviewable comment -K1eYxzEdqaB2ozcA14j:-851265641 -->

---

<sup>**[src/main/CoverageTrack.js, line 185 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/338#-K1eZ5dDWVtE3k2eq93-)** ([raw file](https://github.com/hammerlab/pileup.js/blob/a15c96995339285e5cab206b984f2d9c4cdfb29e/src/main/CoverageTrack.js#L185)):</sup>
nice touch -- thanks!
<!-- Reviewable comment -K1eZ5dDWVtE3k2eq930:1009061367 -->

---

<sup>**[src/main/CoverageTrack.js, line 196 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/338#-K1eZmXi36pOet-7w7v2)** ([raw file](https://github.com/hammerlab/pileup.js/blob/a15c96995339285e5cab206b984f2d9c4cdfb29e/src/main/CoverageTrack.js#L196)):</sup>
do you think it is worth keeping this padding only when the view is zoomed in at a level that you can distinguish individual bases? I don't like the Moiré pattern, but it looks better when you are investigating individual bases. 

Up to you.
<!-- Reviewable comment -K1eZmXjGTai7UgxDexp:908817110 -->

---

<sup>**[src/main/CoverageTrack.js, line 226 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/338#-K1eZAzkSSiTRzmTH0Zz)** ([raw file](https://github.com/hammerlab/pileup.js/blob/a15c96995339285e5cab206b984f2d9c4cdfb29e/src/main/CoverageTrack.js#L226)):</sup>
aha! I didn't know you could strokeText on canvas and that is why I dropped those halos around the text :\ (TIL)
<!-- Reviewable comment -K1eZAzkSSiTRzmTH0_-:1432906468 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pileup.js/338)
<!-- Sent from Reviewable.io -->

danvk commented:
I added the padding back. Will merge on green.
<!-- Reviewable comment -K1eykfjm2NayXLmzQQk:-1782929271 -->

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

<sup>**[src/main/CoverageTrack.js, line 174 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/338#-K1eYxzEdqaB2ozcA14i)** ([raw file](https://github.com/hammerlab/pileup.js/blob/a15c96995339285e5cab206b984f2d9c4cdfb29e/src/main/CoverageTrack.js#L174)):</sup>
See the note on [this MDN page][1]. It's not valid JS syntax without the parens.

This is sort of a long-standing wart in JS... for example, if you type `{}` and hit enter in the Chrome console, it'll spit back `undefined`. I think it's trying to interpret `{}` as a block of code, not an object.

[1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Syntax
<!-- Reviewable comment -K1esddUQmvsVnNGpD7G:-380829473 -->

---

<sup>**[src/main/CoverageTrack.js, line 196 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/338#-K1eZmXi36pOet-7w7v2)** ([raw file](https://github.com/hammerlab/pileup.js/blob/a15c96995339285e5cab206b984f2d9c4cdfb29e/src/main/CoverageTrack.js#L196)):</sup>
Yeah, it definitely looks better with the padding. I added it back & refactored this method a bit.

Here's what it looks like at different resolutions. I rounded all coordinates to get crisp lines. 
![screen shot 2015-10-27 at 4 15 24 pm](https://cloud.githubusercontent.com/assets/98301/10771467/4f3f0e88-7cc7-11e5-922b-844efbbf85db.png)
![screen shot 2015-10-27 at 4 21 32 pm](https://cloud.githubusercontent.com/assets/98301/10771462/4457aa16-7cc7-11e5-94b3-d92cc384d727.png)
![screen shot 2015-10-27 at 4 21 39 pm](https://cloud.githubusercontent.com/assets/98301/10771433/2c9476b6-7cc7-11e5-8c6f-3dc93164384b.png)
<!-- Reviewable comment -K1eyWvpcJi6qO9p8Msr:52761684 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pileup.js/338)
<!-- Sent from Reviewable.io -->

armish commented:
This looks even better! :)
<!-- Reviewable comment -K1f-1rNzQEMzSU6nmDS:1848567483 -->

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[src/main/CoverageTrack.js, line 174 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/338#-K1eYxzEdqaB2ozcA14i)** ([raw file](https://github.com/hammerlab/pileup.js/blob/a15c96995339285e5cab206b984f2d9c4cdfb29e/src/main/CoverageTrack.js#L174)):</sup>
aha, tricky one! Thanks for the link.
<!-- Reviewable comment -K1eziYJUf86urVuJOwY:-552312309 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pileup.js/338)
<!-- Sent from Reviewable.io -->
